### PR TITLE
Service migration script

### PR DESCRIPTION
**Pull request by saurabhvaidvmware**

### CODE REVIEW

File: scripts/service_migration.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    Arguments:
+    v1 :: CoreV1Api object
+    svc :: original service as dictionary
+    --------------
+    Returns:
+    None
+    ************************************************************************
+    '''
+    
+    svc['metadata']['annotations']['load-balancer-ips'] = svc['status']['load_balancer']['ingress'][0]['ip']
```
Comment (apalsule) [inline scripts/service_migration.py:22]: make the annotation a constant defined in script

**Comment by apalsule** [comment]
General comment : get rid of all print statements added for debugging purposes, and only print actual errors. Also print an output of services updated and services not updated.

File: scripts/service_migration.py (hunk 1 above)
Comment (saurabhvaidvmware) [reply scripts/service_migration.py]: constant added to script

**Comment by saurabhvaidvmware** [comment]
> General comment : get rid of all print statements added for debugging purposes, and only print actual errors. Also print an output of services updated and services not updated.

Print statements removed. Added list of Loadbalancer type services updated and not updated.

File: scripts/service_migration.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
+import copy
```
Comment (swathinsankaran) [inline scripts/service_migration.py:1]: please add a copyright license agreement here.

File: scripts/service_migration.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+from kubernetes import client, config
+from kubernetes.client.exceptions import ApiException
+import argparse
+import os.path
+from pprint import pprint
+
+IP_ANNOTATION = 'load-balancer-ips'
+def update_annot(v1, svc):
```
Comment (swathinsankaran) [inline scripts/service_migration.py:9]: update_annotation will be more readable I think.

File: scripts/service_migration.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    Arguments:
+    v1 :: CoreV1Api object
+    svc :: original service as dictionary
+    --------------
+    Returns:
+    None
+    ************************************************************************
+    '''
+    
+    svc['metadata']['annotations'][IP_ANNOTATION] = svc['status']['load_balancer']['ingress'][0]['ip']
```
Comment (swathinsankaran) [inline scripts/service_migration.py:23]: Better to check the presence of keys `status`, `load_balancer`, `ip` and length of svc.

File: scripts/service_migration.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    service = client.V1Service()
+    service.api_version = "v1"
+    service.kind = "Service"
+    service.metadata = svc['metadata']
+    service.spec = svc['spec']
+    service.status = svc['status']
+    try:
+        api_response = v1.patch_namespaced_service(name = svc["metadata"]["name"], namespace="avi-system", body=service)
+        #updated_service = v1.read_namespaced_service(name = svc["metadata"]["name"], namespace="avi-system")
```
Comment (swathinsankaran) [inline scripts/service_migration.py:32]: we can remove this statement if we are not planning to add this in the future.

File: scripts/service_migration.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    except ApiException as e:
+        print(e)
+
+def get_svc(config_file):
```
Comment (swathinsankaran) [inline scripts/service_migration.py:36]: get_svc name is misleading, it does get of services as well as update of annotation. I think a better naming is required here.

File: scripts/service_migration.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+def main():
+    parser = argparse.ArgumentParser(description="Script to read all namespaced services and add external IP to service annotation for LoadBalancer type services")
+    parser.add_argument('kubeconfig_file', help="Path of kubeconfig file in local directory")
+    args = parser.parse_args()
+    config_file = args.kubeconfig_file
+    while not os.path.isfile(config_file):
+        config_file = input("Give valid kubeconfig filepath (Input \'End\' to exit program): ")
+        if config_file == 'End' or config_file=='end':
+            break
+    if os.path.isfile(config_file):
```
Comment (swathinsankaran) [inline scripts/service_migration.py:84]: Are we doing the same condition check in L80?

Comment (swathinsankaran) [reply scripts/service_migration.py:84]: We need it because there is a break statement to end the script.

File: scripts/service_migration.py (hunk 3 above)
Comment (saurabhvaidvmware) [reply scripts/service_migration.py]: Updated function name

File: scripts/service_migration.py (hunk 2 above)
Comment (saurabhvaidvmware) [reply scripts/service_migration.py:1]: Added copyright license agreement

File: scripts/service_migration.py (hunk 5 above)
Comment (saurabhvaidvmware) [reply scripts/service_migration.py]: Removed statement as per recommendation

File: scripts/service_migration.py (hunk 6 above)
Comment (saurabhvaidvmware) [reply scripts/service_migration.py]: Updated function name

File: scripts/service_migration.py (hunk 4 above)
Comment (saurabhvaidvmware) [reply scripts/service_migration.py]: Added check for existence of each key, and whether said keys hold null/empty values

File: scripts/service_migration.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        api_response = v1.patch_namespaced_service(name = svc["metadata"]["name"], namespace="avi-system", body=service)
+        return True
+    except ApiException as e:
+        print(e)
+        return False
+
+def fetch_lbsvc_for_update(config_file):
+    '''
+    ************************************************************************
+    Takes a kubeconfig file and retrieves all services found in all namespaces. If service if of type LoadBalancer, then calls update_annotation() to patch service
```
Comment (swathinsankaran) [inline scripts/service_migration.py:58]: If service `is` of type LoadBalancer.

**Review by swathinsankaran — DISMISSED** [review]
LGTM.

Please add a few scenarios that you have tested, both positive and negative.

**Comment by saurabhvaidvmware** [comment]
I successfully ran the script with the following situations in the cluster:
1. No LoadBalancer type service present in the cluster.
2. LoadBalancer type service present - 
         a. and annotation missing.
         b. and annotation already populated.
         c. but status field empty.

**Review by akshayhavile — APPROVED** [review]
minor nit. LGTM

**Review by swathinsankaran — APPROVED** [review]

**Review by akshayhavile — APPROVED** [review]

**Comment by swathinsankaran** [comment]
build ako

**Review by apalsule — APPROVED** [review]
